storage: optimize `pointSynthesizingIter` for `MVCCGet`

erikgrinaker opened the pull request:
**storage: fix slice extention in `MVCCRangeKeyStack.CloneInto`**

This patch makes `MVCCRangeKeyStack.CloneInto` extend the existing slice
to its full capacity before growing it. Previously, this could build a
slice that was too small and panic.

Release note: None

**storage: add `MVCCIterator.IsPrefix()`**

This method allows detecting whether an iterator is a prefix iterator,
which comes in handy for wrapping iterators to apply optimizations.

Resolves #86104.

Release note: None

**storage: reuse `pointSynthesizingIter` range key slice**

```
name                                                                 old time/op    new time/op    delta
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=0-24     3.70µs ± 1%    3.67µs ± 1%  -0.89%  (p=0.002 n=10+10)
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=1-24     7.37µs ± 0%    7.25µs ± 0%  -1.70%  (p=0.000 n=9+10)
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=10-24    14.7µs ± 1%    14.4µs ± 1%  -2.21%  (p=0.000 n=10+10)
```

Touches #83049.

Release note: None

**storage: add prefix mode for `pointSynthesizingIter`**

This patch adds a `prefix` mode for `pointSynthesizingIter`, detected
from the parent's `IsPrefix()` method. When enabled, this allows
omitting key cloning and comparisons. This replaces the previous
`emitOnSeekGE` parameter.

```
name                                                                 old time/op    new time/op    delta
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=0-24     3.67µs ± 1%    3.67µs ± 1%    ~     (p=0.517 n=10+10)
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=1-24     7.25µs ± 0%    7.07µs ± 0%  -2.45%  (p=0.000 n=10+9)
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=10-24    14.4µs ± 1%    14.2µs ± 1%  -1.10%  (p=0.000 n=10+10)
```

Touches #83049.

Release note: None

erikgrinaker committed:
storage: reuse `pointSynthesizingIter` range key slice

```
name                                                                 old time/op    new time/op    delta
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=0-24     3.70µs ± 1%    3.67µs ± 1%  -0.89%  (p=0.002 n=10+10)
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=1-24     7.37µs ± 0%    7.25µs ± 0%  -1.70%  (p=0.000 n=9+10)
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=10-24    14.7µs ± 1%    14.4µs ± 1%  -2.21%  (p=0.000 n=10+10)
```

Release note: None

erikgrinaker committed:
storage: fix slice extention in `MVCCRangeKeyStack.CloneInto`

This patch makes `MVCCRangeKeyStack.CloneInto` extend the existing slice
to its full capacity before growing it. Previously, this could build a
slice that was too small and panic.

Release note: None

erikgrinaker committed:
storage: add `prefix` mode for `pointSynthesizingIter`

This patch adds a `prefix` mode for `pointSynthesizingIter`, detected
from the parent's `IsPrefix()` method. When enabled, this allows
omitting key cloning and comparisons. This replaces the previous
`emitOnSeekGE` parameter.

```
name                                                                 old time/op    new time/op    delta
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=0-24     3.67µs ± 1%    3.67µs ± 1%    ~     (p=0.517 n=10+10)
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=1-24     7.25µs ± 0%    7.07µs ± 0%  -2.45%  (p=0.000 n=10+9)
MVCCGet_Pebble/batch=true/versions=1/valueSize=8/numRangeKeys=10-24    14.4µs ± 1%    14.2µs ± 1%  -1.10%  (p=0.000 n=10+10)
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86093)
<!-- Reviewable:end -->

erikgrinaker committed:
storage: add `MVCCIterator.IsPrefix()`

This method allows detecting whether an iterator is a prefix iterator,
which comes in handy for wrapping iterators to apply optimizations.

Release note: None

nicktrav reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 6 of 6 files at r2, 1 of 1 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86093#-:-N9SnnS4BqHi1g6MdS0B:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jbowens)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
TFTR! CI failures are known flake.

bors r=nicktrav
